sandboxes/dynamicanalysis/analyze-python.py

```diff
@@ -24,54 +32,216 @@ def install_arg(self) -> str:
+            if should_investigate(return_type):
+                print("[investigate type]", return_type)
+                mark_seen(return_type)
+                try_call_methods(return_value, return_type, should_investigate, mark_seen)
+        elif inspect.isclass(member):
+            instance = try_instantiate_class(member, name)
+            assert instance.__class__ == member
+            if instance is not None and member not in instantiated_types:
+                instantiated_types.add(member)
+                try_call_methods(instance, name, should_investigate, mark_seen)
+        else:
+            skipped_names.append(name)
+
+    print("[skipped members]", " ".join(skipped_names))
+
+
+def alarm_handler(sig_num, frame):
+    raise TimeoutError("Timeout exceeded for function execution")
+
+
+# Call a function with mock arguments based on its declared signature.
+# The arguments are of type MagicMock, whose instances will return
+# dummy values for any method called on them.
+# Exceptions must be handled by the caller.
+def invoke_function(obj):
+    signature = inspect.signature(obj)
+    args = []
+    kwargs = {}
+
+    for name, param in signature.parameters.items():
+        # use MagicMock to create semi-realistic function argument values
+        # https://docs.python.org/3/library/unittest.mock.html
+        value = MagicMock() if param.default == param.empty else param.default
+
+        match param.kind:
+            case param.POSITIONAL_ONLY:
+                args.append(value)
+            case param.KEYWORD_ONLY | param.POSITIONAL_OR_KEYWORD:
+                kwargs[name] = value
+            case param.VAR_POSITIONAL:  # when *args appears in signature
+                pass  # ignore
+            case param.VAR_KEYWORD:  # when **args appears in signature
+                pass  # ignore
+
+    # bind args and invoke the function
+    # any exceptions will be propagated to the caller
+    bound = signature.bind(*args, **kwargs)
+
+    # run with timeout to prevent hangs
+    signal.alarm(EXECUTION_TIMEOUT_SECONDS)
+    ret = obj(*bound.args, **bound.kwargs)
+    signal.alarm(0)
+    return ret
+
+
+# Execute a callable and catch any exception, logging to stdout
+def run_and_catch_all(c: callable):
+    try:
+        return c()
+    except BaseException as e:
+        # catch ALL exceptions, including KeyboardInterrupt and system exit
+        print(type(e), e, sep=": ")
+
+
+def try_invoke_function(f, name, is_method=False):
+    print("[method]" if is_method else "[function]", name)
+
+    def invoke():
+        return invoke_function(f)
+
+    ret = run_and_catch_all(invoke)
+
+    if ret is not None:
+        print("[return value]", repr(ret))
+        return ret
+
+
+def try_instantiate_class(c, name):
+    print("[class]", name)
+
+    def instantiate():
+        return invoke_function(c)
+
+    return run_and_catch_all(instantiate)
+
+
+def try_call_methods(instance, class_name, should_investigate, mark_seen):
```

Comment:
nit: should we add a TODO to handle running async python code? Assuming we need to handle them differently.

Reply:
Done